README.md
# Markdown Mail

Command `>mdm: Send markdown to mail` sends markdown file to mail.

## Front Matter

### Own mail server

This follows YAML front matter.
For example, use double quotes to force a string.

```markdown
---
host: mail.example.com
port: 25
from: from@example.com
to: to@example.com
subject: "[This] is email subject"
---

# Markdown Mail
```

### Gmail

[Turn on "Less secure app access"](https://myaccount.google.com/lesssecureapps).
This setting is not available for accounts with 2-Step Verification enabled.

> [Less secure apps & your Google Account](https://support.google.com/accounts/answer/6010255?hl=en#zippy=%2Cif-less-secure-app-access-is-on-for-your-account) - Google Account Help

```markdown
---
host: smtp.gmail.com
port: 587
from: from@gmail.com
to: to@example.com
subject: "[This] is email subject"
---

# Markdown Mail
```

## Contributing

```sh
code mdm
```

```sh
rm -rf node_modules out
```

```sh
npm install
```

```sh
npm run pretest
```

## Testing

```sh
npm test
```

In VS Code, press `F5` to open a new window with your extension loaded.

## References

- [Extension API](https://code.visualstudio.com/api)
- [Your First Extention](https://code.visualstudio.com/api/get-started/your-first-extension)
- [Extension Anatomy](https://code.visualstudio.com/api/get-started/extension-anatomy)
- [Extension Guides](https://code.visualstudio.com/api/extension-guides/overview)
- [Publishing Extensions](https://code.visualstudio.com/api/working-with-extensions/publishing-extension)
- [What is the Simple Mail Transfer Protocol (SMTP)?](https://www.cloudflare.com/learning/email-security/what-is-smtp/)
  - [ko-kr](https://www.cloudflare.com/ko-kr/learning/email-security/what-is-smtp/)


CHANGELOG.md
# Change Log

## v0.2.0

- Support Gmail using "Less secure app access"

## v0.1.1

- Fix [#1](https://github.com/markruler/mdm/issues/1)

## v0.1.0

- Sending email using own email server


src/test/suite/reader.test.ts
import * as assert from "assert";
import { UserOption } from "../../extension";

import { readMailOptions } from "../../reader";

suite("Reader Test Suite", () => {
  test("Read front matter", async () => {
    const option: UserOption = await readMailOptions(`---
host: mail.example.com
port: 25
from: from@example.com
to: to@example.com
subject: "[This] is email subject"
---

# This is email body
`);
    assert.strictEqual(option.host, "mail.example.com");
    assert.strictEqual(option.port, 25);
    assert.strictEqual(option.from, "from@example.com");
    assert.strictEqual(option.to, "to@example.com");
    assert.strictEqual(option.subject, "[This] is email subject");
  });
});


src/mailer.ts
import * as vscode from "vscode";
import { SendMailOptions, createTransport } from "nodemailer";
import { UserOption } from "./extension";

/**
 * Send mail
 * @param body Mail body
 * @param userOption User options
 */
export async function sendMail(body: string, userOption: UserOption) {
  let password;
  if (userOption.host === "smtp.gmail.com") {
    password = await vscode.window.showInputBox({
      placeHolder: "********",
      prompt: "Gmail password",
      password: true,
    });
  }

  const transporter = createTransport({
    host: userOption.host,
    port: userOption.port,
    secure: userOption.port === 465 ? true : false,
    auth: {
      user: userOption.from,
      pass: password,
    }
  });

  const options: SendMailOptions = { ...userOption, html: body };

  await transporter.sendMail(options).then((info) => {
    console.log("Message sent: %s", info.response);
  });
}

/**
 * Return mail server from user input or configuration.
 * @returns mailServer (host:port)
 */
export async function getMailServer() {
  let mailServer = await vscode.window.showInputBox({
    placeHolder: "host:port",
    prompt: "Set mail server (host:port)",
  });

  if (!mailServer) {
    mailServer = vscode.workspace.getConfiguration("mdm").get("mailServer");
  }

  if (!mailServer) {
    throw new Error("mailServer is undefined");
  }

  return mailServer;
}


src/extension.ts
import * as vscode from "vscode";
import * as MarkdownIt from "markdown-it";
import { getMailServer, sendMail } from "./mailer";
import { read, readMailOptions } from "./reader";

export interface UserOption {
  host: string; // mail.example.com
  port: number; // 25, 465, 587
  from: string; // from@example.com
  to: string; // to@example.com
  subject: string; // Title
}

/**
 * Send markdown to mail server
 */
async function sendMarkdown() {
  const markdown: MarkdownIt = new MarkdownIt().use(
    // Remove front matter
    require("markdown-it-front-matter"),
    function (frontMatter: string) {
      console.log(frontMatter);
    }
  );

  if (!vscode.window.activeTextEditor) {
    return; // no editor
  }

  const { document } = vscode.window.activeTextEditor;

  const scheme = document.uri.scheme;
  if (!scheme || scheme !== "file") {
    return; // not a file
  }
  console.log(`current file ${document.uri}`);

  const originalText: string = (await read(document.uri)) || "";
  const userOption: UserOption = await readMailOptions(originalText);
  console.log("User Option", userOption);

  if (!userOption.host || !userOption.port) {
    const mailServer: string = await getMailServer();
    console.log(`mailServer: ${mailServer}`);

    const mailHostPort = mailServer.split(":");
    userOption.host = mailHostPort[0];
    userOption.port = parseInt(mailHostPort[1]);
  }

  const html = markdown.render(originalText);

  await sendMail(html, userOption);

  vscode.window.showInformationMessage(
    `Send to ${userOption.host}:${userOption.port}`
  );
}

/**
 * Activate extension
 * @param context Extension context
 */
export async function activate(context: vscode.ExtensionContext) {
  console.log("Activate MDM");

  let send = vscode.commands.registerCommand("mdm.send", () => sendMarkdown());
  context.subscriptions.push(send);
}

/**
 * Deactivate extension
 */
export function deactivate() {
  console.log("Deactivate MDM");
}
